tests/run_data_utils_test.py

```diff
@@ -0,0 +1,53 @@
+from batchglm.api.utils.data import constraint_system_from_star
+import pandas as pd
+import logging
+import numpy as np
+import unittest
+
+logger = logging.getLogger("batchglm")
+
+class TestConstraintSystemFromStar(unittest.TestCase):
+
+    def test_constraint_system_dict(self):
+        formula = '~0 + condition + batch'
+        sample_description = pd.DataFrame({ 'condition': [0, 1, 1, 0, 0, 1], 'batch': [1, 2, 2, 0, 0, 3] })
+        constraints = {"batch": "condition"}
+        dmat, coef_names, cmat, term_names = constraint_system_from_star(constraints, sample_description=sample_description, formula=formula)
+        assert term_names == ['condition', 'batch']
+        assert coef_names == ['condition[0]', 'condition[1]', 'batch[0]', 'batch[1]', 'batch[2]', 'batch[3]']
+        assert np.all(np.equal(cmat, np.array([[ 1.,  0.,  0.,  0.],
+            [ 0.,  1.,  0.,  0.],
+            [ 0.,  0., -1.,  0.],
+            [ 0.,  0.,  1.,  0.],
+            [ 0.,  0.,  0., -1.],
+            [ 0.,  0.,  0.,  1.]])))
+        assert np.all(np.equal(dmat, np.array([[1., 0., 0., 1., 0., 0.],
+            [0., 1., 0., 0., 1., 0.],
+            [0., 1., 0., 0., 1., 0.],
+            [1., 0., 1., 0., 0., 0.],
+            [1., 0., 1., 0., 0., 0.],
+            [0., 1., 0., 0., 0., 1.]])))
+    
+    def test_constraint_system_list(self):
+        formula = '~0 + condition + batch'
+        sample_description = pd.DataFrame({ 'condition': [0, 1, 1, 0, 0, 1], 'batch': [1, 2, 2, 0, 0, 3] })
+        constraints = ["condition[0] + condition[1] = 0"]
+        dmat, coef_names, cmat, term_names = constraint_system_from_star(constraints, sample_description=sample_description, formula=formula)
+        assert term_names == None
+        assert coef_names == ['condition[0]', 'condition[1]', 'batch[T.1]', 'batch[T.2]', 'batch[T.3]']
+        assert np.all(np.equal(cmat, np.array([[-1.,  0.,  0.,  0.],
+                [ 1.,  0.,  0.,  0.],
+                [ 0.,  1.,  0.,  0.],
+                [ 0.,  0.,  1.,  0.],
+                [ 0.,  0.,  0.,  1.]])))
+        assert np.all(np.equal(dmat, np. array([[1., 0., 1., 0., 0.],
+                [0., 1., 0., 1., 0.],
+                [0., 1., 0., 1., 0.],
+                [1., 0., 0., 0., 0.],
+                [1., 0., 0., 0., 0.],
+                [0., 1., 0., 0., 1.]])))
```

Comment:
@davidsebfischer @picciama shouldn't this test cast's true design matrix be identical to the one above?

Reply:
See comment above: The dmat above is having the coeffs (cond0, cond1, batch0/1/2/3). The below one it constrained differently. Here it seems the last three coeffs correspond to batch 1/2/3 while the first two are the two conditions. You can imagine that the batch 0 can be reflected by simply taking condition 0  and subtracting the batch 1. This is indeed what happens if you have a look at the constraints matrix where `constraints[0][0] = -1`. This means that the resulting linear combination after applying `np.matmul(dmat, cmat)` results in this matrix:
```py
[[-1.  1.  0.  0.]
 [ 1.  0.  1.  0.]
 [ 1.  0.  1.  0.]
 [-1.  0.  0.  0.]
 [-1.  0.  0.  0.]
 [ 1.  0.  0.  1.]]

``` 

This makes sense: The4th and 5th row represent the samples from batch 0. The 1st row is batch 1. So this is a system that represents batches 1, 2 and 3 with respect to `- batch 0 + condition1`. So 1st column is the intercept. You can now subtract the intercept which results in `- (-batch 0 + condition1) = batch 0 - condition1 = batch 0` or add it which results in `-batch 0 + condition 1 = condition 1` which you can use in combination with other coeffs to get to the desired batch that only contains samples with condition 1.

Reply:
More Info: In my opinion, it isn't really a problem to have a different dmat. This one is the `np.matmul(dmat, cmat)` from the constraint system using the dict `{'batch': 'condition'}`:
```py
[[ 1.  0.  1.  0.]
 [ 0.  1.  0. -1.]
 [ 0.  1.  0. -1.]
 [ 1.  0. -1.  0.]
 [ 1.  0. -1.  0.]
 [ 0.  1.  0.  1.]]

```
I find this more difficult to interpret. I think the 1st column encode `condition 0 + midpoint` and the 2nd one `condition 1 +midpoint 2` where `midpoint = abs(batch 0-batch1) / 2` and `midpoint 2 = abs(batch 2-batch3) / 2` The 3rd column is accordingly `abs(batch 0-batch1) / 2` which means adding and subtracting this will lead you towards batch 1 and 2 respectively. Same thing for 4th column for batch 2/3.

Not sure if this is correct but it's the only thing that makes sense to me given the constraints matrix.  @davidsebfischer is this even desirable? We have no intercept here it seems.

Reply:
Right @picciama this was my point - I understand why it's doing this, but the behavior should be consistent IMO.  If the design matrix is just an "untransformed" representation, it shouldn't change based on the constraints which do the transformation IMO.  I understand dropping a coefficient since it is "unecessary" but the behavior should be consistent and I think is related to my other comment about the error.